src/devtools/client/debugger/src/actions/pause/paused.js

```diff
@@ -47,32 +62,45 @@ function fetchAsyncFrames(cx) {
  * @static
  */
 export function paused({ executionPoint, time }) {
-  return async function (dispatch, getState) {
-    dispatch({ type: "PAUSED", executionPoint, time });
+  return async function (dispatch, getState, { ThreadFront }) {
+    const state = getState();
     trackEvent("paused");
 
-    // Get a context capturing the newly paused and selected thread.
-    const cx = getThreadContext(getState());
+    await dispatch(pauseRequestedAt(executionPoint, time));
+
+
+    const pause = ThreadFront.ensurePause(executionPoint, time);
+
+    dispatch({ type: "PAUSED", executionPoint, time, id: pause.id });
+
+    try {
+      await pause.createWaiter;
+    } catch (e) {
+      console.error(e);
+      dispatch(failedToCreatePause(pause.id));
+    }
+
+    console.log({ pause, waiter: pause.createWaiter });
 
-    await dispatch(fetchFrames(cx));
+    await dispatch(fetchFrames(pause));
 
-    const frame = getSelectedFrame(getState());
+    const frame = getSelectedFrame(state);
```

Comment:
:hand: Bit of a difference in semantics with this change. We went from reading the state _right now_, to reusing the state from _earlier_ at the start of the thunk, before we awaited anything. May not be critical, but _could_ be meaningful.  Just checking to see if that's what we want here.

Reply:
This is something that *really* bugs me about this code. There's a lot of

```
setLocalState()
dispatch(loadAdditionalState())
...
// in the thunk
readTheLocalState()
fetchAdditionalDataBasedOnLocalState()
...
```

It all just assumes that you can dispatch actions and the right thing will happen because you mutated some state right before dispatching. It would be a *lot* clearer and easier to follow if it was like `dispatch(loadAdditionalState(localState))`. Anyways, all that to say that yes, this *is* an intentional change. Right before we dispatch `paused` we set `ThreadFront.currentExecutionPoint` and `ThreadFront.currentTime`, and then when we call dispatch `fetchFrames` we tell `ThreadFront` to go create a pause for its `currentExecutionPoint` and `currentTime` (and we hope those are the same as they were at the beginning of the action). I don't think it's wise to change this all in one fell swoop, but I can at least try and make it behave a little more predictably 🤷

Reply:
This pattern/abiguity is another type of thing that would be nice to address when we tackle the architecture.